internal/entitlement/connector.go

```diff
@@ -159,26 +160,82 @@ func (c *entitlementConnector) CreateEntitlement(ctx context.Context, input Crea
 			return nil, err
 		}
 
-		return ent, nil
-	})
+		return ent, err
+	}
 
-	return ent, err
+	if ctxTx, err := entutils.GetTxDriver(ctx); err == nil {
+		// we're already in a tx
+		return doInTx(ctx, ctxTx)
+	} else {
+		return entutils.StartAndRunTx(ctx, c.entitlementRepo, doInTx)
+	}
+}
+
+func (c *entitlementConnector) OverrideEntitlement(ctx context.Context, subject string, entitlementIdOrFeatureKey string, input CreateEntitlementInputs) (*Entitlement, error) {
+	// Validate input
+	if subject != input.SubjectKey {
+		return nil, &models.GenericUserError{Message: "Subject key in path and body do not match"}
+	}
+
+	oldEnt, err := c.entitlementRepo.GetEntitlementOfSubject(ctx, input.Namespace, input.SubjectKey, entitlementIdOrFeatureKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if oldEnt == nil {
+		return nil, fmt.Errorf("inconsistency error, entitlement not found: %s", entitlementIdOrFeatureKey)
+	}
+
+	if oldEnt.DeletedAt != nil {
+		return nil, fmt.Errorf("inconsistency error, entitlement already deleted: %s", oldEnt.ID)
+	}
+
+	// ID has priority over key
+	featureIdOrKey := input.FeatureID
+	if featureIdOrKey == nil {
+		featureIdOrKey = input.FeatureKey
+	}
+	if featureIdOrKey == nil {
+		return nil, &models.GenericUserError{Message: "Feature ID or Key is required"}
+	}
+
+	feature, err := c.featureConnector.GetFeature(ctx, input.Namespace, *featureIdOrKey, productcatalog.IncludeArchivedFeatureFalse)
+	if err != nil || feature == nil {
+		return nil, &productcatalog.FeatureNotFoundError{ID: *featureIdOrKey}
+	}
+
+	if feature.ID != oldEnt.FeatureID {
+		return nil, &models.GenericUserError{Message: "Feature in path and body do not match"}
+	}
+
+	// Do the override in TX
+	return entutils.StartAndRunTx(ctx, c.entitlementRepo, func(ctx context.Context, tx *entutils.TxDriver) (*Entitlement, error) {
+		ctx = entutils.NewTxContext(ctx, tx)
+
+		// Delete previous entitlement
+		// FIXME: we publish an event during this even if we fail later
```

Comment:
This should be fine for now, but I agree that we should revamp trns handling, and provide hooks for publishing only if the whole transaction seems ok to commit (it's not 100% but better than the current situation)